### ISSUE TITLE
ASN1_OBJECT_new can be deprecated and removed

### ISSUE THREAD

**Issue by davidben** [openssl/openssl#16862]
`ASN1_OBJECT_new` is documented as [almost never used](https://www.openssl.org/docs/manmaster/man3/ASN1_OBJECT_new.html). This is because it is practically useless outside OpenSSL internals. It returns an empty `ASN1_OBJECT`, but the struct is opaque and there are no setters in the public API to mutate an `ASN1_OBJECT`, which means it can never be filled in. Instead, `ASN1_OBJECT`s are constructed with all their parameters, using functions like `ASN1_OBJECT_create` or `d2i_ASN1_OBJECT`, or looked up by NID with `OBJ_nid2obj` and friends.

Thus we may as well deprecate it and remove it in a later release.

**Comment by fwh-dc** [comment]
@nhorman @t8m 
Found this one. Is it required that we do deprecation before 4.0 to be able to remove it at 5.0? And is there a guide on how we mark functions for deprecation?

**Comment by t8m** [comment]
I do not think there is a guide. It would be certainly nice to have one. It is fairly easy - just look for example at b2168cd81db how the EVP_MD_CTX_get0_md_data() function was deprecated there.

**Comment by t8m** [comment]
> Is it required that we do deprecation before 4.0 to be able to remove it at 5.0?

Yeah. Unless there is a very good reason for an exception, it should be done like this.

**Comment by davidben** [comment]
Doesn't matter much either way, but I will note it is basically impossible for a caller to do anything useful with this function. :-) It creates an empty `ASN1_OBJECT` that will forever stay in the empty state, because OpenSSL (rightfully) does not expose any setters for `ASN1_OBJECT`.
